kingpin/actors/aws/ecs.py

```
@@ -406,7 +420,369 @@ def _execute(self):
+            service_name: name of the service to wait for.
+        """
+        # Used to ensure event logs are only printed once.
+        seen_events = set()
+        while True:
+            done = yield self._service_deployed(
+                service_name,
+                seen_events)
+            if done:
+                break
+            yield gen.sleep(10)
+
+    @gen.coroutine
+    def _service_deployed(self, service_name, seen_events):
+        """Checks if service is finished deploying.
+
+        Args:
+            service_name: name of the service to wait for.
+            seen_events: set of seen event ids.
+
+        Returns:
+            A boolean indicating whether the service is completely deployed.
+        """
+        service = yield self._describe_service(service_name)
+
+        deployments = service['deployments']
+        primary_deployment = None
+        for deployment in deployments:
+            if deployment['status'] == 'PRIMARY':
+                primary_deployment = deployment
+                break
+        if not primary_deployment:
+            # There should always be one 'PRIMARY' deployment returned.
+            raise exceptions.UnrecoverableActorFailure(
+                'No primary deployment.')
+
+        service_timestamp = primary_deployment['createdAt']
+
+        sorted_new_events = _get_sorted_new_log_events(
+            events=service['events'],
+            start_timestamp=service_timestamp,
+            seen_events=seen_events)
+
+        for event in sorted_new_events:
+            event_timestamp, event_message = event
+            self.log.info('Event: {message}'.format(
+                message=event_message))
+
+        running_count = primary_deployment['runningCount']
+        desired_count = primary_deployment['desiredCount']
+        missing_count = desired_count - running_count
+
+        extra_deployment_count = len(deployments) - 1
+
+        if missing_count == 0 and extra_deployment_count == 0:
```

Good call. Let's just throw an exception in that case for now and see if it comes up.